diff -Naru alsa-lib-1.0.26.orig/src/conf/cards/Makefile.am alsa-lib-1.0.26/src/conf/cards/Makefile.am
--- alsa-lib-1.0.26.orig/src/conf/cards/Makefile.am	2016-09-21 20:01:01.744931616 +0800
+++ alsa-lib-1.0.26/src/conf/cards/Makefile.am	2016-09-21 20:01:26.728595098 +0800
@@ -1,60 +1,6 @@
 alsaconfigdir = @ALSA_CONFIG_DIR@
 alsadir = $(alsaconfigdir)/cards
-cfg_files = aliases.conf \
-	AACI.conf \
-	ATIIXP.conf \
-	ATIIXP-SPDMA.conf \
-	ATIIXP-MODEM.conf \
-	AU8810.conf \
-	AU8820.conf \
-	AU8830.conf \
-	Audigy.conf \
-	Audigy2.conf \
-	Aureon51.conf \
-	Aureon71.conf \
-	CA0106.conf \
-	CMI8338.conf \
-	CMI8338-SWIEC.conf \
-	CMI8738-MC6.conf \
-	CMI8738-MC8.conf \
-	CMI8788.conf \
-	CS46xx.conf \
-	Echo_Echo3G.conf \
-	EMU10K1.conf \
-	EMU10K1X.conf \
-	ENS1370.conf \
-	ENS1371.conf \
-	ES1968.conf \
-	FM801.conf \
-	FWSpeakers.conf \
-	FireWave.conf \
-	GUS.conf \
-	HDA-Intel.conf \
-	ICE1712.conf \
-	ICE1724.conf \
-	ICH.conf \
-	ICH4.conf \
-	ICH-MODEM.conf \
-	Maestro3.conf \
-	NFORCE.conf \
-	PC-Speaker.conf \
-	PMac.conf \
-	PMacToonie.conf \
-	PS3.conf \
-	RME9636.conf \
-	RME9652.conf \
-	SI7018.conf \
-	SB-XFi.conf \
-	TRID4DWAVENX.conf \
-	USB-Audio.conf \
-	YMF744.conf \
-	VIA686A.conf \
-	VIA8233.conf \
-	VIA8233A.conf \
-	VIA8237.conf \
-	VX222.conf \
-	VXPocket.conf \
-	VXPocket440.conf
+cfg_files = aliases.conf 
 
 if BUILD_ALISP
 cfg_files += aliases.alisp
diff -Naru alsa-lib-1.0.26.orig/src/conf/pcm/Makefile.am alsa-lib-1.0.26/src/conf/pcm/Makefile.am
--- alsa-lib-1.0.26.orig/src/conf/pcm/Makefile.am	2016-09-21 20:01:01.760931401 +0800
+++ alsa-lib-1.0.26/src/conf/pcm/Makefile.am	2016-09-21 20:01:52.944242034 +0800
@@ -1,9 +1,4 @@
-cfg_files = default.conf front.conf rear.conf center_lfe.conf side.conf\
-	    surround40.conf surround41.conf \
-	    surround50.conf surround51.conf \
-	    surround71.conf iec958.conf hdmi.conf modem.conf \
-	    dmix.conf dsnoop.conf \
-	    dpl.conf
+cfg_files = default.conf dmix.conf dsnoop.conf 
 
 EXTRA_DIST = $(cfg_files)
 
